wsmux/session.go

```diff
@@ -0,0 +1,351 @@
+
+	close(s.closed)
+	return err
+}
+
+// Addr used for implementing net.Listener
+func (s *Session) Addr() net.Addr {
+	return s.conn.LocalAddr()
+}
+
+func (s *Session) closeHandler(code int, text string) error {
+	s.logger.Printf("ws conn closed: code %d : %s", code, text)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.closeConn = false
+	return s.Close()
+}
+
+func (s *Session) recvLoop() {
+	for {
+		select {
+		case <-s.closed:
+			return
+		default:
+		}
+
+		t, msg, err := s.conn.ReadMessage()
+		if err != nil {
+			_ = s.abort(err)
+			break
+		}
+		if t != websocket.BinaryMessage {
+			s.logger.Print("did not receive binary message")
+			continue
+		}
+
+		if len(msg) < 5 {
+			s.logger.Print(errMalformedHeader)
+			_ = s.abort(errMalformedHeader)
+		}
+
+		h := header(msg[:5])
+		msg = msg[5:]
+
+		id, msgType := h.id(), h.msg()
+
+		switch msgType {
+		//Used for creating a new stream
+		case msgSYN:
+			s.mu.Lock()
+			if _, ok := s.streams[id]; ok {
+				s.logger.Printf("received duplicate syn frame for stream %d", id)
+				s.mu.Unlock()
+				break
+			}
+
+			str := newStream(id, s)
+			// no point in locking here
+			str.accept(DefaultCapacity)
+
+			s.streams[id] = str
+			if err := s.send(newAckFrame(id, uint32(DefaultCapacity))); err != nil {
+				s.logger.Printf("error accepting stream %d", id)
+				_ = s.Close()
+				return
+			}
+			s.asyncPushStream(str)
+			s.mu.Unlock()
+
+		//received data
+		case msgDAT:
+			s.mu.Lock()
+			str, ok := s.streams[id]
+			s.mu.Unlock()
+			if !ok {
+				s.logger.Printf("received data frame for unknown stream %d", id)
+				break
+			}
+			str.push(msg)
+			s.logger.Printf("received DAT frame on stream %d: %v", id, bytes.NewBuffer(msg))
+
+		//received ack frame
+		case msgACK:
+			s.mu.Lock()
+			str, ok := s.streams[id]
+			s.mu.Unlock()
+			s.logger.Printf("received ack frame")
+			if !ok {
+				s.logger.Printf("received ack frame for unknown stream %d", id)
+				break
+			}
+
+			read := binary.LittleEndian.Uint32(msg)
+			select {
+			case <-str.accepted:
+				s.logger.Printf("received ack frame: id %d: remote read %d bytes", id, read)
+				str.unblock(read)
+			default:
+				// close str.accepted to accept stream
+				s.logger.Printf("accepting stream")
+				str.accept(read)
+				break
+			}
+
+		// received fin frame
+		case msgFIN:
+			s.mu.Lock()
+			str, ok := s.streams[id]
+			s.mu.Unlock()
+			if !ok {
+				s.logger.Printf("received fin frame for unknown stream %d", id)
+				break
+			}
+
+			str.setRemoteClosed()
+		}
+
+	}
+}
+
+func (s *Session) removeStream(id uint32) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if _, ok := s.streams[id]; ok {
+		delete(s.streams, id)
+	}
+}
+
+func (s *Session) asyncPushStream(str *stream) {
+	select {
+	case s.streamCh <- str:
+	default:
+	}
+}
+
+func (s *Session) abort(e error) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.acceptErr = e
+
+	if s.IsClosed() {
+		return nil
+	}
+
+	return s.Close()
+}
+
+func (s *Session) removeDeadStreams() {
```

Comment:
I'm this function does something... but I don't know what... it disparately needs a comment explaining it...